### PR TITLE
Parse `using` declaration (explicit resource management)

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Implement parsing/printing support of the stage-2 [Explicit Resource Management](https://github.com/tc39/proposal-explicit-resource-management) proposal
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | https://github.com/babel/website/pull/2684
| Any Dependency Changes?  |
| License                  | MIT

This PR parses / generates the using declaration
```js
{
  using handler = getHandler()
}
```
introduced by the Explicit Resource Management proposal. We also create the new syntax plugin and add it to the babel-standalone.

I suggest reviewing this PR by commits.

/cc @rbuckton

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14968"></a>

### CODE REVIEW

File: packages/babel-parser/src/parser/statement.ts [hunk 1]
```diff
@@ -314,6 +314,42 @@ export default abstract class StatementParser extends ExpressionParser {
...
+      const originalState = this.state;
+      // @ts-expect-error: Second lookahead
+      this.state = lookahead;
+      const lookaheadAhead = this.lookahead();
+      this.state = originalState;
+      if (
+        lookaheadAhead.type === tt._of &&
+        lookaheadAhead.end - lookaheadAhead.start === 2
+      ) {
+        // using of of ...
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/statement.ts:330]: In the worst scenario, we have to do triple lookahead to differentiate between `using of of of` and `using of of`.

@rbuckton Maybe we can forbid `of` as a using binding in `for-of`? So `for (using of` must be a for-LHS and we don't have to peek more.

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/statement.ts:330]: Isn't it the same as `let of of of`? Or is `of` disallowed in that case?

Comment (JLHwung) [reply packages/babel-parser/src/parser/statement.ts:330]: > for ( [lookahead ∉ { let, async of }] [LeftHandSideExpression](https://tc39.es/ecma262/#prod-LeftHandSideExpression)[?Yield, ?Await] of [AssignmentExpression](https://tc39.es/ecma262/#prod-AssignmentExpression)[+In, ?Yield, ?Await] ) [Statement](https://tc39.es/ecma262/#prod-Statement)[?Yield, ?Await, ?Return]

for-of has disabled `let` and `async` as LHS so `for(let of of` must start a for-of with a let-binding `of`. But `using` comes after for-of so we have to recognize `for(using of of` as for-of LHS.

File: packages/babel-parser/src/parser/statement.ts [hunk 2]
```diff
@@ -314,6 +314,42 @@ export default abstract class StatementParser extends ExpressionParser {
...
+        const nextNextNextTokenStart = this.nextTokenStartSince(
+          lookaheadAhead.end,
+        );
+        if (
+          this.input.charCodeAt(nextNextNextTokenStart) ===
+          charCodes.rightParenthesis
```
Comment (JLHwung) [inline packages/babel-parser/src/parser/statement.ts:336]: The check here does not work when the for-of-right is a TSAsExpression:
```ts
for (using of of as T);
```

File: packages/babel-parser/src/parser/statement.ts (hunk 1 above)
Comment (JLHwung) [reply packages/babel-parser/src/parser/statement.ts:330]: Per https://github.com/tc39/proposal-explicit-resource-management/issues/107, I will disallow `of` as a for-using-of binding name, hopefully we will get rid of extra lookahead, too.

File: packages/babel-generator/src/generators/statements.ts [hunk 3]
```diff
@@ -303,17 +311,27 @@ export function VariableDeclaration(
...
 
   //
 
-  this.printList(node.declarations, node, { separator });
+  let iterator: PrintJoinOptions["iterator"] | undefined;
+  if (kind === "using") {
+    // Ensure no line break between `using` and the first declarator
+    iterator = (_, i: number) => {
+      if (i === 0) {
+        this._noLineTerminator = _noLineTerminator;
```
Comment (nicolo-ribaudo) [inline packages/babel-generator/src/generators/statements.ts:329]: Shouldn't we also unset `_noLineTerminator` after printing?

File: packages/babel-parser/src/parser/statement.ts [hunk 4]
```diff
@@ -314,6 +314,17 @@ export default abstract class StatementParser extends ExpressionParser {
     return false;
   }
 
+  startsUsingForOf(): boolean {
+    const lookahead = this.lookahead();
+    if (lookahead.type === tt._of && lookahead.end - lookahead.start === 2) {
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/statement.ts:319]: Can you add a comment saying that we check that the length is 2 because of escape sequences?

**Review by nicolo-ribaudo — APPROVED** [review]

File: packages/babel-parser/src/parser/statement.ts (hunk 1 above)
Comment (rbuckton) [reply packages/babel-parser/src/parser/statement.ts]: There is now a [PR](https://github.com/tc39/proposal-explicit-resource-management/pull/108) that addresses this in the proposal repository.

File: packages/babel-generator/src/generators/statements.ts (hunk 3 above)
Comment (JLHwung) [reply packages/babel-generator/src/generators/statements.ts:301]: Do you mean unset `_noLineTerminator` after printing the `using` keyword? The variable `_noLineTerminator` here stores the value of `this.noLineTerminator` before we print `using`, and then we restore `this._noLineTerminator` after the first declarator is printed.

File: packages/babel-parser/src/parser/statement.ts (hunk 4 above)
Comment (JLHwung) [reply packages/babel-parser/src/parser/statement.ts]: I end up using the `containsEsc` flag.

File: packages/babel-generator/test/fixtures/using-declarations/basic/output.js [hunk 5]
```diff
@@ -1,6 +1,4 @@
 {
   using /* 1 */a = foo(),
-
-  /* 2 */
-  b = foo();
+    /* 2 */b = foo();
 }
```
Comment (liuxingbaoyu) [inline packages/babel-generator/test/fixtures/using-declarations/basic/output.js:4]: The comments look good.😃

**Review by liuxingbaoyu — APPROVED** [review]
Just a small question: I noticed that the newly added `containsEsc` property is only doing reads and no writes, is this expected?

**Comment by JLHwung** [comment]
> Just a small question: I noticed that the newly added `containsEsc` property is only doing reads and no writes, is this expected?

The `containsEsc` will be set by the tokenizer by `readWord`

https://github.com/babel/babel/blob/0b0f083c777bf59b197467b973d6418325c5ddd7/packages/babel-parser/src/tokenizer/index.ts#L1380

So if we see an identifier, `lookahead.containsEsc` will be a boolean, otherwise `lookahead` will not have a `containsEsc` property.

**Pull request merged**